src/vmm/src/vstate/memory.rs

```diff
@@ -283,20 +306,49 @@ impl GuestMemoryExtension for GuestMemoryMmap {
         let mut writer_offset = 0;
         let page_size = get_page_size().map_err(MemoryError::PageSize)?;
 
+        // Merging KVM dirty pages bitmap into internal firecracker bitmap. Needed to keep track of
+        // KVM dirty pages in case we need to use them again if current dump operation fails.
         self.iter()
             .enumerate()
             .try_for_each(|(slot, region)| {
+                let firecracker_bitmap = region.bitmap();
                 let kvm_bitmap = dirty_bitmap.get(&slot).unwrap();
+                for (i, v) in kvm_bitmap.iter().enumerate() {
+                    // FIXME: looks like this loop can be replaced with some
+                    // `(kvm_dirty_pages | fc_dirty_pages)`
+                    for j in 0..64 {
+                        let is_kvm_page_dirty = ((v >> j) & 1u64) != 0u64;
+                        let page_offset = ((i * 64) + j) * page_size;
+
+                        if is_kvm_page_dirty {
+                            // FIXME: not sure about passing page_size as len
+                            firecracker_bitmap.mark_dirty(page_offset, page_size)
+                        }
+                    }
+                }
+
+                Ok(())
+            })
+            .map_err(MemoryError::WriteMemory);
+
+        self.iter()
+            .enumerate()
+            .try_for_each(|(slot, region)| {
                 let firecracker_bitmap = region.bitmap();
+                let kvm_bitmap = dirty_bitmap.get(&slot).unwrap();
```

Comment:
do we need this at all? can't we just iterate over  firecracker_bitmap?

Reply:
Yes, we don't need KVM bitmap after we merged it into the FC one.